test/es6/iteratorclose.js

```diff
@@ -0,0 +1,950 @@
+            }
+            
+            [a] = gf();
+            assert.areEqual(finallyCount, 1);
+            assert.areEqual(a, 1);
+            
+            finallyCount = 0;
+            assert.throws(function () { [obj.prop] = gf(); }, Error, "Setting value will throw", "From setter");
+            assert.areEqual(finallyCount, 1);
+            
+            function* gf2() {
+                yield 1;
+                assert.fail('should not reach after yield');
+            }
+            var returnCount = 0;
+            gf2.prototype.return = function() {
+                returnCount++;
+                return {};
+            };
+            [a] = gf2();
+            assert.areEqual(returnCount, 1);
+            
+            gf2.prototype.return = function() {
+                returnCount++;
+                throw new Error('Exception from return function');
+            };
+            assert.throws(function () { [a] = gf2(); }, Error, "Return function throws", "Exception from return function");
+            returnCount = 0;
+            assert.throws(function () { [obj.prop] = gf2(); }, Error, "Exception at destructuring element wins", "From setter");
+            assert.areEqual(returnCount, 1);
+		}
+	},    
+    {
+		name : "Destructuring - at function parameter",
+		body : function () {
+            nextCount = 0;
+            returnCount = 0;
+            (function([a,b]) {})(iterable);
+            assert.areEqual(nextCount, 2);
+            assert.areEqual(returnCount, 1);
+            
+            shouldReturnThrow = true;
+            assert.throws(function () { (function([a,b]) {})(iterable) }, Error, "Calling return function while at param throws", "Exception from return function");
+            shouldReturnThrow = false;
+		}
+	},
+    {
+		name : "Destructuring - assigning to rest parameter can throw but should not call .return function",
+		body : function () {
+            function* gf() {
+                yield 1;
+                yield 2;
+            }
+            gf.prototype.return = function() {
+                assert.fail('return function should not be called');
+            };
+            
+            assert.throws(function () { [...obj.prop] = gf(); }, Error, "Assigning to rest head can throw", "From setter");
+		}
+	},    
+    {
+		name : "Destructuring - .next throws during rest assignment but it should not call the .return function",
+		body : function () {
+            returnCount = 0;
+            shouldNextThrow = true;
+            assert.throws(function () { var [...a] = iterable; }, Error, "Assigning to rest head can throw", "Exception from next function");
+            assert.throws(function () { [...a] = iterable; }, Error, "Assigning to rest head can throw", "Exception from next function");
+            shouldNextThrow = false;
+            assert.areEqual(returnCount, 0);
+		}
+	},    
+    {
+		name : "For..of - validation of calling .return function on abrupt loop break",

```

Comment:
test262 has some good test cases for the iteratorclose. I'll report once I get the result on them.